added button to reshuffle encounter discard pile into deck

dscarpac opened the pull request:
XML button provides functionality

Entrox-Licher commented on src/core/MythosArea.ttslua:
```diff
@@ -201,6 +201,27 @@ function reshuffleEncounterDeck(params)
+  if isReshuffling then return end
+  isReshuffling = true
+
+  -- flip deck and then shuffle
```
You can just combine the deck objects directly using the `putObject` [function](https://api.tabletopsimulator.com/object/#putobject).  It should drastically simplify the logic here.

dscarpac replied:
Yeah, @Chr1Z93 has already mentioned this. I'm going to work on this a bit more and clean up the other encounter draw functions in the same script so they all match.

Re: nicheness -- the other use case is when the encounter deck runs out on its own. It only reshuffles when you draw an encounter card, but the encounter deck has to reshuffle after the last card is drawn and resolved. So, e.g. if the last card drawn is an enemy, if you defeat it, it goes into the discard after the encounter deck has already reshuffled. If there's another spot for it, I'm all ears, but IMO it's an important enough function to not just be a hotkey.

Entrox-Licher replied:
Oh wow that's super interesting, I didn't know that's how the rules worked!  Definitely been playing that one wrong XD. 
 Too much muscle memory from other card games (even this one!  Horror/reshuffle of player cards only triggers on the next instance of draw, not when empty.  What an odd deviation).  In this case, I think it would be better to just change the existing script to check if the encounter deck is empty _after_ drawing, rather than before, and trigger the reshuffle, in accordance to the rules, as normal there.  

Then that leaves this button with just the first niche use case, which I don't think justifies its existence (as a button, hotkey or anything else).  We have very specifically strayed away from automating scenario-specific effects (aside from the odd set-up helper), so I don't really see a reason to start now, especially for such a basic action.

dscarpac replied:
I may have communicated this poorly -- the encounter card draw script definitely should stay as it is -- we can't predict when a card has finished resolving, so the encounter deck reshuffle has to be triggered manually.

Entrox-Licher replied:
Aaaaaah I see what you're saying now.  Hmmm I'm starting to come around to the use case then.  I still think we want to add 2 broadcast messages then: one when the deck is empty, to remind people to click the button to reshuffle, and one if the old reshuffle script is triggered, to 'chastise' them for it a bit and make sure they're playing it correctly.  Or maybe even disable the default reshuffle altogether honestly, to avoid that play pattern trap.

That being said, if it is going to be much more frequently called, I would consider also making a hotkey for it.  I'd let @Chr1Z93 weigh in on that decision though.

dscarpac replied:
I'm also rewriting some of the rest of the MythosArea script under @Chr1Z93's tutelage -- some of this script makes parts of the other scripts redundant, whether or not we have a button to call it. You'll see some more soon :-D

Entrox-Licher reviewed (dismissed):
Should clean up the function a bit there.  

I also wonder how much this feature is really needed?  Since the encounter discard already shuffles automatically if empty when drawing encounter cards, the only real use case for this is when specifically instructed to shuffle the encounter discard into the deck (which on most scenarios will happen once, maybe twice?).  Seems a bit niche to merit a spot on the mat

Glitchy-Tozier commented:
I do think this comes up at least once or twice each scenario (mostly when the agenda flips).

_However_, i also agree that this functionality might not deserve a spot. Currently, space is reserved for functionality that does one of the following:
+ comes up very often and saves you a small amout of work (discard encounter card)
+ comes up seldomly but saves you a lot of work (remove all doom/clues from play)
+ (comes up often and saves you a lot of work (upkeep button))

This new button seems to be in the fourth category: It comes up rarely AND saves you only a minimal amout of work.

Chr1Z93 commented:
1) Imo a button to reshuffle the discard pile would be useful enough to merit its existence. Agendas (or even acts) ask you to do this at least every other game (probably way more on average).

2) As @dscarpac correctly pointed out, the reshuffling when drawing a new card is too late. So I do like this addition. There should probably be a reminder somewhere to trigger this (maybe directly on the MythosArea, below the encounter deck?).

Chr1Z93 commented on src/core/MythosArea.ttslua:
```diff
@@ -160,7 +154,14 @@ function drawEncounterCard(params)
-    reshuffleEncounterDeck(params)
+    if reshuffledAlready == true then
+      reshuffledAlready = false
+      -- print something?
```
Should replace this with an actual message 🤣

Chr1Z93 commented on src/core/MythosArea.ttslua:
```diff
@@ -180,27 +181,31 @@ function actualEncounterCardDraw(card, params)
-function reshuffleEncounterDeck(params)
-  -- flag to avoid multiple calls
+function reshuffleEncounterDeck()
+  --flag to avoid multiple calls
```
Nitpick: Don't remove the space

Chr1Z93 commented on src/core/MythosArea.ttslua:
```diff
@@ -180,27 +181,31 @@ function actualEncounterCardDraw(card, params)
+    
+  if #discardPile > 0 then
+    local discardDeck = discardPile[1]
+    if not discardDeck.is_face_down then discardDeck.setRotation({0, -90, 180}) end --flips discard pile
```
Nitpick: Split this into 3 lines

Chr1Z93 commented on src/core/MythosArea.ttslua:
```diff
@@ -180,27 +181,31 @@ function actualEncounterCardDraw(card, params)
+      encounterDeck.putObject(discardDeck)
+      encounterDeck.shuffle()
+    end
+      broadcastToAll("Shuffled encounter discard into deck.", "White")
```
Nitpick: Fix indentation

Chr1Z93 commented on xml/MythosArea.xml:
```diff
@@ -0,0 +1,10 @@
+<Panel 
+  position="124 40 -13"
+  rotation="0 0 180"
+  height="74" width="74">
+
+
+    <Button scale="0.1 0.1 1"  fontsize="62" onClick="reshuffleEncounterDeck">➡</Button>
+
+
+</Panel>
```
Formatting is all over the place. I can recommend this VS Code extension for formatting: https://marketplace.visualstudio.com/items?itemName=DotJoshJohnson.xml

dscarpac commented:
Cleaned up the button, added mouseover and buttonpress colors:
![image](https://github.com/argonui/SCED/assets/148364064/665210f9-df86-47df-8bd4-df5bb4cfc7f8)
![image](https://github.com/argonui/SCED/assets/148364064/51397b2d-5c30-40ba-9390-e07b09f4c453)

Chr1Z93 reviewed (approved):
looks good to me!
